rsx/rsx_lib_gl.cpp

```diff
@@ -990,126 +990,94 @@ static void GlRenderer_draw(GlRenderer *renderer)
 
    renderer->command_buffer->map = NULL;
 
-   GLushort *opaque_triangle_indices =
-      renderer->opaque_triangle_indices + renderer->opaque_triangle_index_pos + 1;
-   GLsizei opaque_triangle_len =
-      INDEX_BUFFER_LEN - renderer->opaque_triangle_index_pos - 1;
+   if (!renderer->batches.empty())
+	  renderer->batches.back().count = renderer->vertex_index_pos
+		 - renderer->batches.back().first;
 
-   if (opaque_triangle_len)
+   for (std::vector<PrimitiveBatch>::iterator it =
+		  renderer->batches.begin();
+		  it != renderer->batches.end();
+		  ++it)
    {
-      if (!DRAWBUFFER_IS_EMPTY(renderer->command_buffer))
-      {
-         /* This method doesn't call prepare_draw/finalize_draw itself, it
-          * must be handled by the caller. This is because this command
-          * can be called several times on the same buffer (i.e. multiple
-          * draw calls between the prepare/finalize) */
-         glDrawElements(GL_TRIANGLES, opaque_triangle_len, GL_UNSIGNED_SHORT, opaque_triangle_indices);
-      }
+	  /* Mask bits */
+	  if (it->set_mask)
+		 glStencilOp(GL_KEEP, GL_KEEP, GL_REPLACE);
+	  else
+		 glStencilOp(GL_KEEP, GL_KEEP, GL_ZERO);
```

Comment:
I think this is kind of dangerous with textured, non-transparent primitives. You will end up filling the stencil buffer with either 0 or 1 depending on the alpha bit of the texture. You could use stencil_export, but that's not widely supported.

Reply:
Is the alpha in the framebuffer kept in sync with the stencil bit with texturing, copies, etc?

Reply:
For textured opaque primitives, only the fully transparent texels are discarded (0x0000) by the shader and will not modify the stencil. So I think it's ok (?)

No, the stencil is not kept in sync for copies. However it is kept in sync (reset) in fill_rect. It *should* be kept in sync with primitives based on the gpu mask bits.